packages/dumb/src/components/settings/history/history-table-content.tsx

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2023, Diabeloop
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import _ from 'lodash'
+import { HistorySpannedRow } from './history-table-spanned-row'
+import { HistoryTableRow } from './history-table-row'
+import React, { FunctionComponent } from 'react'
+import { getLongDayHourFormat } from '../../../utils/datetime/datetime.util'
+import moment from 'moment-timezone'
+import { TimePrefs, Unit } from 'medical-domain'
+import { ChangeType, HistorizedParameter, IncomingRow } from '../../../models/historized-parameter.model'
+
+interface HistoryTableContentProps {
+  onSwitchToDaily: Function
+  data: IncomingRow[]
+  timePrefs: TimePrefs
+  length: number
+}
+
+export const HistoryTableContent: FunctionComponent<HistoryTableContentProps> = (props): JSX.Element => {
+  const { onSwitchToDaily, data, timePrefs, length } = props
+
+  // TODO: Candidate to custom hooks ?
+  const buildAllRows = (history: IncomingRow[]): HistorizedParameter[] => {
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches to further communicate your intent.  -->
**🎯 suggestion (blocking):** ‏Some improvements that could be done to enhance the readability of this method.

- Invert the `if ()` statements and use early returns
Example:
```ts
if (!_.isArray(history)) {
  return xxx
}
// Rest of the code
```

- Given that we parse an array of `IncomingRow`s and we want to output another array (of a different size) of `HistorizedParameter`s, the whole logic could be done in a `reduce()`
```ts
const rows = history.reduce((currentRows: HistorizedParameter[], historyElement: IncomingRow) => {
  // Logic goes here
}, [])
```

- Given that we want 1 row per `parameter` from `historyElement.parameters`, we could define a method `buildRowFromParameter()` and use a `map()` instead of the `for ()` loop
```ts
const buildRowFromParameter = (parameter: xxx): HistorizedParameter => {
  const row: HistorizedParameter = { ...parameter } as HistorizedParameter
  row.rawData = parameter.name

  // Rest of the logic here

  return row
}

// Call it within `buildAllRows()` method
const parameterRows = parameters.map((parameter: Parameter2) => getRowByParameter(parameter))
```

- The `latestDate` could be retrieved via its own method (using a `reduce()`) instead of being defined and updated in the `for ()` loop

- The `mLatestDate` could be defined via a ternary condition instead of a `let`

- The `currentParameters` logic could be handled in a dedicated method

I have it all locally and it looks like it's working, don't hesitate to ask me if this can help 😄

Reply:
i hope i forget nothing, if it is the case ping me @celia-rouis 😉